### ISSUE TITLE
Implementation is leaking exception data

### ISSUE THREAD

**Issue by devinbost** [prometheus/client_golang#781]
We have an implementation of Prometheus in our Apache Pulsar framework, and the current Java implementation is leaking memory. Every time an exception is thrown, the exception is appended. As a consequence, the metrics scrapes continue to grow in size. One of the metrics scrapes was reported as ~100 MB. However, the Java implementation of Prometheus was outdated, so I followed Prometheus guidelines when re-writing it for our Go implementation. However, it's not clear if this implementation solves the leaking memory issue either. 

Here is what that looks like:

```
systemExceptions = prometheus.NewGaugeVec(
	prometheus.GaugeOpts{
		Name: PulsarFunctionMetricsPrefix + "system_exception",
		Help: "Exception from system code."}, exceptionMetricsLabelNames)
. . . 

func (stat *StatWithLabelValues) reportSystemExceptionPrometheus(exception error, ts int64) {
	errorTs := []string{exception.Error(), strconv.FormatInt(ts, 10)}
	exceptionMetricLabels := append(stat.metricsLabels, errorTs...)
	systemExceptions.WithLabelValues(exceptionMetricLabels...).Set(1.0)
}
```
(https://github.com/apache/pulsar/blob/master/pulsar-function-go/pf/stats.go#L261)

The issue was reported here: https://github.com/apache/pulsar/issues/7489

What are the recommendations for including function exceptions when we report them through a Gague? 
How do we avoid these memory problems?

**Comment by beorn7** [comment]
tl;dr: The Go implementation will “leak” memory, too. That's by design.

Longer answer:

Each time you call `systemExceptions.WithLabelValues(exceptionMetricLabels...).Set(1.0)`, you are using a new unique combination of label values (if I understand your code snippet correctly). This will inevitably let the memory usage of the `GaugeVec` grow forever, as it remembers every one of those. And so will the exposition size grow forever. The reason is that Prometheus instrumentation is for metrics, and metrics are expected to be set or incremented many times in their life cycle. Here, every unique element in the `GaugeVec` is set exactly once. What you are doing is essentially event logging. And Prometheus is not the right system for that. For this use case, you would log the exception and feed the logs to a logs processing system. (If you want to use a logs processing system that is "like Prometheus, just for logs", I can recommend [Loki](https://grafana.com/oss/loki/). Disclaimer: It's an OSS project lead by my employer and started by my boss. ;o)

For further discussion about how to use Prometheus (and how not to), I'd like to refer you to the [prometheus-users mailing list](https://groups.google.com/forum/#!forum/prometheus-users). On the mailing list, more people are available to potentially respond to your question, and the whole community can benefit from the answers provided.
